server/src/main/java/org/apache/druid/discovery/LocalDruidNodeDiscoveryProvider.java

```diff
@@ -0,0 +1,164 @@
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BooleanSupplier;
+
+/**
+ * Local node discovery. In this mode, Druid may run a number of
+ * logical servers within a single process, so that node discovery
+ * is all within the same process, but matches a true distributed
+ * node discovery.
+ * <p>
+ * Concurrency is at the level of the entire provider, which is
+ * necessary to coordinate the active state with ongoing node
+ * discovery and listener registrations. Such a "global" lock
+ * is fine because the actions don't occur frequently, nor do the
+ * actions take much time. For this same reason, we use plain old
+ * Java synchronization rather than a fancier locking mechanism.
+ * <p>
+ * At present, this class is used in unit tests so that it is
+ * possible to simulate cluster membership changes without actually running
+ * ZK, etc. The class can become the primary provider if/when Druid
+ * can run all services in a single process: in that case, we won't
+ * need ZK to tell the in-process services about each other.
+ */
+public class LocalDruidNodeDiscoveryProvider extends DruidNodeDiscoveryProvider implements DruidNodeAnnouncer
+{
+  private class RoleEntry implements DruidNodeDiscovery
+  {
+    private final Map<DruidNode, DiscoveryDruidNode> nodes = new HashMap<>();
+    private final List<Listener> listeners = new ArrayList<>();
+
+    private void register(DiscoveryDruidNode node)
+    {
+      List<DruidNodeDiscovery.Listener> targets;
+      synchronized (LocalDruidNodeDiscoveryProvider.this) {
+        DiscoveryDruidNode prev = nodes.put(node.getDruidNode(), node);
+        if (prev != null) {
+          return;
+        }
+        targets = new ArrayList<>(listeners);
+      }
+      for (Listener listener : targets) {
+        listener.nodesAdded(Collections.singletonList(node));
+      }
+    }
+
+    private void deregister(DiscoveryDruidNode node)
+    {
+      List<DruidNodeDiscovery.Listener> targets;
+      synchronized (LocalDruidNodeDiscoveryProvider.this) {
+        DiscoveryDruidNode prev = nodes.remove(node.getDruidNode());
+        if (prev == null) {
+          return;
+        }
+        targets = new ArrayList<>(listeners);
+      }
+      for (Listener listener : targets) {
+        listener.nodesRemoved(Collections.singletonList(node));
+      }
+    }
+
+    @Override
+    public Collection<DiscoveryDruidNode> getAllNodes()
+    {
+      synchronized (LocalDruidNodeDiscoveryProvider.this) {
+        return new ArrayList<>(nodes.values());
+      }
+    }
+
+    @Override
+    public void registerListener(Listener listener)
+    {
+      synchronized (LocalDruidNodeDiscoveryProvider.this) {
+        listeners.add(listener);
+        if (!active) {
+          return;
+        }
+      }
+      listener.nodeViewInitialized();
+    }
+
+    private synchronized boolean contains(DruidNode node)
```

Comment:
Does this method need to be synchronized on both `RoleEntry.this` and `LocalDruidNodeDiscoveryProvider.this`?

Reply:
Good catch. No, it doesn't. The `LocalDruidNodeDiscoveryProvider` is the sync point. This isn't going to be a high-frequency API, so locking the main object is fine.